core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java

```diff
@@ -0,0 +1,180 @@
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.MutableShardRouting;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.gateway.AsyncShardFetch;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * Transport action that reads the cluster state for any unassigned shards for specific indices
+ * and fetches the shard version, host node id and any exception while opening the shard index
+ * from all the nodes.
+ */
+public class TransportIndicesUnassignedShardsAction extends TransportMasterNodeReadAction<IndicesUnassignedShardsRequest, IndicesUnassigedShardsResponse> {
+
+    private final TransportNodesListGatewayStartedShards listShardsInfo;
+
+    @Inject
+    public TransportIndicesUnassignedShardsAction(Settings settings, TransportService transportService, ClusterService clusterService, ThreadPool threadPool, ActionFilters actionFilters,
+                                                  TransportNodesListGatewayStartedShards listShardsInfo) {
+        super(settings, IndicesUnassignedShardsAction.NAME, transportService, clusterService, threadPool, actionFilters, IndicesUnassignedShardsRequest.class);
+        this.listShardsInfo = listShardsInfo;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected IndicesUnassigedShardsResponse newResponse() {
+        return new IndicesUnassigedShardsResponse();
+    }
+
+    @Override
+    protected void masterOperation(IndicesUnassignedShardsRequest request, ClusterState state, ActionListener<IndicesUnassigedShardsResponse> listener) {
+        // collect all unassigned shards for the requested indices
+        final DiscoveryNodes nodes = state.nodes();
+        final MetaData metaData = state.metaData();
+        final RoutingNodes routingNodes = state.routingNodes();
+        Set<String> requestedIndices = new HashSet<>();
+        requestedIndices.addAll(Arrays.asList(request.indices()));
+        List<ShardId> shardIdsToFetch = new ArrayList<>();
+        for (MutableShardRouting shard : Iterables.concat(routingNodes.unassigned(), routingNodes.ignoredUnassigned())) {
```

Comment:
changed to using `state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)`